modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/io/StatsDReporter.scala

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (c) 2021 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.enrich.fs2.io
+
+import cats.syntax.show._
+
+import java.net.{DatagramPacket, DatagramSocket, InetAddress}
+import java.nio.charset.StandardCharsets.UTF_8
+
+import cats.effect.{Blocker, ContextShift, Resource, Sync, Timer => CatsTimer}
+import fs2.{Pure, Stream}
+import com.codahale.metrics._
+import scala.jdk.CollectionConverters._
+import scala.concurrent.duration.FiniteDuration
+
+import _root_.io.chrisdavenport.log4cats.Logger
+import _root_.io.chrisdavenport.log4cats.slf4j.Slf4jLogger
+
+import com.snowplowanalytics.snowplow.enrich.fs2.config.io.MetricsReporter
+
+/**
+ * Reports metrics to a StatsD server over UDP
+ *
+ * We use the DogStatsD extension to the StatsD protocol, which adds arbitrary key-value tags to the metric, e.g:
+ * `snowplow.enrich.good.count:20000|g|#app_id:12345,env:prod`
+ */
+object StatsDReporter {
+
+  /**
+   * A stream which periodically sends metrics from the registry to the StatsD server.
+   *
+   * The stream refreshes the InetAddress of the server each time it sends metrics, instead of
+   * caching the inet address, or keeping a long-running channel open. This allows us to
```

Comment:
Can you confirm that even though we are calling `InetAddress.getByName()` each time we want to send metrics, there is some caching and the DNS request is performed only when the TTL has expired ?It would be great to add it in the comment.

Reply:
The details are in [the InetAddress docs](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/net/InetAddress.html) in the section starting "InetAddress Caching".  The important line is

> The default setting is to cache for an implementation specific period of time.

Meaning it depends on our JVM, and on security settings.  I have tested this locally and I found it resolved a new IP address very quickly.  But in the version we deploy to kubernetes we might find it behaves differently - I have no idea.  We need to test it.

I can put some of these details in the comment, as you suggest.

Reply:
I found the answer in the `java.security` file of our docker image:

```
# The Java-level namelookup cache policy for successful lookups:
#
# any negative value: caching forever
# any positive value: the number of seconds to cache an address for
# zero: do not cache
#
# default value is forever (FOREVER). For security reasons, this
# caching is made forever when a security manager is set. When a security
# manager is not set, the default behavior in this implementation
# is to cache for 30 seconds.
#
# NOTE: setting this to anything other than the default value can have
#       serious security implications. Do not set it unless
#       you are sure you are not exposed to DNS spoofing attack.
#
#networkaddress.cache.ttl=-1
```

So for us it is 30 seconds.

Reply:
Awesome, thank you for the details!